common/src/Model/Brush.cpp

```diff
@@ -916,80 +936,95 @@ namespace TrenchBroom {
-                return false;
+            if (!allowVertexRemoval) {
+                // All moving vertices must still be present in the result
+                for (const Vec3& movingVertex : moving.vertexPositions()) {
+                    if (!result.hasVertex(movingVertex + delta))
+                        return CanMoveVerticesResult::rejectVertexMove();
+                }
+            }
             
             // Will the result go out of world bounds?
             if (!worldBounds.contains(result.bounds()))
-                return false;
+                return CanMoveVerticesResult::rejectVertexMove();
             
             // Will the brush become invalid?
             if (!result.polyhedron())
-                return false;
+                return CanMoveVerticesResult::rejectVertexMove();
             
             // One of the remaining two ok cases?
             if ((moving.point() && remaining.polygon()) ||
                 (moving.edge() && remaining.edge()))
-                return true;
+                return CanMoveVerticesResult::acceptVertexMove(result);
             
             // Invert if necessary.
             if (remaining.point() || remaining.edge() || (remaining.polygon() && moving.polyhedron())) {
                 using std::swap;
                 swap(remaining, moving);
                 delta = -delta;
             }
-            
-            // Now check if any of the moving vertices would travel into or through the remaining fragment.
+
+            // Now check if any of the moving vertices would travel through the remaining fragment and out the other side.
             for (const BrushVertex* vertex : moving.vertices()) {
                 const Vec3& oldPos = vertex->position();
                 const Vec3 newPos = oldPos + delta;
-                
+
                 for (const BrushFaceGeometry* face : remaining.faces()) {
-                    if (face->pointStatus(oldPos) == Math::PointStatus::PSAbove &&
-                        face->pointStatus(newPos) == Math::PointStatus::PSBelow) {
+                    if (face->pointStatus(oldPos) == Math::PointStatus::PSBelow &&
+                        face->pointStatus(newPos) == Math::PointStatus::PSAbove) {
                         const Ray3 ray(oldPos, (newPos - oldPos).normalized());
-                        const FloatType distance = face->intersectWithRay(ray, Math::Side_Front);
+                        const FloatType distance = face->intersectWithRay(ray, Math::Side_Back);
                         if (!Math::isnan(distance)) {
                             const FloatType distance2 = distance * distance;
-                            if (distance2 <= ray.squaredDistanceToPoint(newPos).rayDistance)
-                                return false;
+                            const FloatType oldToNew2 = (newPos - oldPos).squaredLength();
+                            if (distance2 <= oldToNew2)
```

Comment:
This check is equal to the condition that the vertex has traveled farther than the distance it had to travel until it hit the face's boundary plane.

Unless I'm not mistaken, this will now always be true, so I think you can just return `CanMoveVerticesResult::rejectVertexMove()` once you find that `distance` is not `NaN`. Sorry I missed this before.